### PR TITLE
[hold/DNM] allow extensions to provide custom Upstreams

### PR DESCRIPTION
**Pull request by stevenctl**
# Description

We have an extension point for upstreams and endpoints with `DiscoveryPlugin`, but the EDS emitting logic is based on HybridUpstreamClient. This adds the ability to provide extension there as well. 

## Code changes
- NewHybridUpstreamClient now takes a list of `[]plugin.Plugin` and adds to the clientMap
- Util function to filter plugins by type using generics
- Replace existing type-filter for DiscoveryPlugin to use helper

## Testing steps

WIP enterprise PoC

# Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: projects/gloo/pkg/plugins/plugins.go [hunk 1]
```diff
@@ -212,6 +212,18 @@ type PluginRegistry interface {
 // It is executed each translation loop, ensuring we have up to date configuration of all plugins
 type PluginRegistryFactory func(ctx context.Context) PluginRegistry
 
+// FilterPlugins attempts to cast each plugin to the given type
+// and returns only elements that cast successfully.
+func FilterPlugins[T Plugin](plugins []Plugin) []T {
```
Comment (nfuden) [inline projects/gloo/pkg/plugins/plugins.go:217]: Nice use of generics!

**Review by timflannagan — APPROVED** [review]

File: projects/gloo/pkg/plugins/plugins.go [hunk 2]
```diff
@@ -212,6 +212,18 @@ type PluginRegistry interface {
...
+	var out []T
```
Comment (timflannagan) [inline projects/gloo/pkg/plugins/plugins.go:218]: prealloc with len(plugins)?

Comment (nfuden) [reply projects/gloo/pkg/plugins/plugins.go:218]: nit: please use the nit prefix per developer guidelines for these types of nonblocking comments

Comment (stevenctl) [reply projects/gloo/pkg/plugins/plugins.go:218]: This really isn't hotpath code (only run at startup, N is small), I'd prefer simplicity legibility over that optimization.

**Comment by sam-heilbron** [comment]
/kick build-bot never started https://storage.googleapis.com/solo-public-build-logs/logs.html?buildid=ba838d71-ecdb-41ed-9bbd-3f6f1f5d1aa5

File: projects/gloo/pkg/upstreams/hybrid_client.go [hunk 3]
```diff
@@ -54,6 +64,17 @@ func NewHybridUpstreamClient(
 		clientMap[sourceConsul] = consul.NewConsulUpstreamClient(consulClient, settings.GetConsulDiscovery())
 	}
 
+	for _, plugin := range clientPlugins {
+		name, client := plugin.SourceName(), plugin.Client()
+		if _, f := clientMap[name]; f {
+			return nil, eris.Errorf("HybridUpstreamClient already has a source %q", name)
```
Comment (sam-heilbron) [inline projects/gloo/pkg/upstreams/hybrid_client.go:70]: If this error occurs, is that a developer error that we could/should error more loudly about, or is this a case that can occur for some customers?

**Review by sam-heilbron — APPROVED** [review]

File: projects/gloo/pkg/upstreams/hybrid_client.go (hunk 3 above)
Comment (stevenctl) [reply projects/gloo/pkg/upstreams/hybrid_client.go:70]: Definitely a dev error. This is called from `RunGlooWithExtensions` which also bubbles up the error, so I think it's crashing, right?

**Comment by stevenctl** [comment]
I might not want to merge this after all, some work @lgadban is doing may make it obsolete.

**Comment by timflannagan** [comment]
@stevenctl is this still relevant based on your last comment or should we close this out?

**Comment by stevenctl** [comment]
closed in favor of KRT

**Pull request closed without merging**